apollo-router/src/test_harness/http_snapshot.rs

```diff
@@ -0,0 +1,563 @@
+                        path: Some(path),
+                        body: request_json_body,
+                    },
+                    response: Response {
+                        status: parts.status.as_u16(),
+                        headers: map_headers(parts.headers, |name| {
+                            state
+                                .include_headers
+                                .as_ref()
+                                .map(|headers| headers.contains(&name.to_string()))
+                                .unwrap_or(true)
+                        }),
+                        body: response_json_body,
+                    },
+                };
+                {
+                    let mut snapshots = state.snapshots.lock().unwrap();
+                    snapshots.insert(snapshot.key(), snapshot.clone());
+                    if let Err(e) = save(state.snapshot_file, &mut snapshots) {
+                        error!(
+                            url = %uri,
+                            method = %method,
+                            error = ?e,
+                            "Unable to save snapshot"
+                        );
+                    }
+                }
+                return Ok(snapshot.try_into().unwrap());
+            } else {
+                let message = "Unable to parse response body as JSON".to_string();
+                error_message = message.clone();
+                error!(
+                    url = %uri,
+                    method = %method,
+                    message,
+                )
+            }
+        } else {
+            let message = "Unable to read response body".to_string();
+            error_message = message.clone();
+            error!(
+                url = %uri,
+                method = %method,
+                message
+            )
+        }
+    }
+
+    Ok(http::Response::builder()
+        .status(500)
+        .body(json!({ "error": error_message}).to_string().into())
+        .unwrap())
+}
+
+fn map_headers<F: Fn(&str) -> bool>(
+    headers: HeaderMap<HeaderValue>,
+    include: F,
+) -> IndexMap<String, Vec<String>> {
+    headers.iter().fold(
+        IndexMap::new(),
+        |mut map: IndexMap<String, Vec<String>>, (name, value)| {
+            let name = name.to_string();
+            if include(&name) {
+                let value = value.to_str().unwrap_or_default().to_string();
+                map.entry(name).or_default().push(value);
+            }
+            map
+        },
+    )
+}
+
+fn save<P: AsRef<Path>>(
+    path: P,
+    snapshots: &mut HashMap<String, Snapshot>,
+) -> Result<(), SnapshotError> {
+    let path = path.as_ref();
+    if let Some(parent) = path.parent() {
+        std::fs::create_dir_all(parent)?;
+    }
+
+    // Keep the snapshots sorted to make source control merges easier when updating snapshots
+    let mut snapshots = snapshots.values().cloned().collect::<Vec<_>>();
+    snapshots.sort_by_key(Snapshot::key);
```

Comment:
If we need to always have the entries sorted by key, then maybe a [`BTreeMap`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html) would work better here since it will always keep its elements sorted by key.

Reply:
Excellent point, no idea why this did not occur to me. I'll update this.